logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/SyncManager.kt

```diff
@@ -0,0 +1,39 @@
+package com.wire.kalium.logic.sync
+
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.flow.updateAndGet
+
+enum class SyncState {
+    WAITING,
+    SLOW_SYNC,
+    COMPLETED,
+}
+
+class SyncManager(private val workScheduler: WorkScheduler) {
+
+    private val internalSyncState = MutableStateFlow(SyncState.WAITING)
+
+    fun completeSlowSync() {
+        internalSyncState.update { SyncState.COMPLETED }
+    }
+
+    suspend fun waitForSlowSyncToComplete() {
+        startSlowSyncIfNotAlreadyCompletedOrRunning()
+        internalSyncState.first { it == SyncState.COMPLETED }
+    }
+
+    private fun startSlowSyncIfNotAlreadyCompletedOrRunning() {
+        val syncState = internalSyncState.updateAndGet {
+            when (it) {
+                SyncState.WAITING -> SyncState.SLOW_SYNC
+                else -> it
+            }
+        }
+
+        if (syncState == SyncState.SLOW_SYNC) {
+            workScheduler.schedule(SlowSyncWorker::class, SlowSyncWorker.name)
```

Comment:
Android's Workers won't necessarily start immediately. 
Maybe we could try to run things without scheduling, and scheduling for the future only in case of failure.

Maybe the `WorkScheduler` can change a bit depending on the platform. It's a pain in the ass.

Nothing I would fix or change now. Just to keep in mind.
As we proceed with Android Reloaded we can fine tune this.

Reply:
What about configuring the work tasks as expedited?https://developer.android.com/topic/libraries/architecture/workmanager/how-to/define-work#expedited

Reply:
It does improve yes. 

I guess we can do that for now without any costs. But the lack of `100% confidence` that the work will start now is always mind bogging.